llpc/util/llpcElfWriter.cpp

```diff
@@ -735,6 +737,285 @@ template <class Elf> void ElfWriter<Elf>::updateMetaNote(Context *pContext, cons
   pNewNote->data = data;
 }
 
+// =====================================================================================================================
+// Get the count of relocs in the reloc section
+//
+// @param pRelocSection : Reloc section
+template <class Elf> size_t ElfWriter<Elf>::getRelocsCount(const SectionBuffer *pRelocSection) {
+  size_t relocCount = 0;
+  if (pRelocSection->secHead.sh_size > 0) {
+    relocCount = static_cast<unsigned>(pRelocSection->secHead.sh_size / pRelocSection->secHead.sh_entsize);
+  }
+
+  return relocCount;
+}
+
+// =====================================================================================================================
+// Get the start pos of relocs used by PS in the reloc section
+//
+// @param pRelocSection : Reloc section
+// @param psIsaOffset : Ps ISA offset in the elf bin
+template <class Elf> size_t ElfWriter<Elf>::getRelocPsStartPos(const SectionBuffer *pRelocSection, size_t psIsaOffset) {
+  size_t psStartPos = 0;
+
+  size_t relocCount = getRelocsCount(pRelocSection);
+  if (relocCount > 0) {
+    auto relocs = reinterpret_cast<typename Elf::Reloc *>(const_cast<uint8_t *>(pRelocSection->data));
+    while (psStartPos < relocCount) {
+      if (relocs[psStartPos].r_offset >= psIsaOffset)
+        break;
+      psStartPos++;
+    }
+  }
+
+  return psStartPos;
+}
+
+// =====================================================================================================================
+// Create a new empty section based on the input section
+//
+// @param [in] pName : Name of the new section
+// @param [in] pInputSection : the input section
+// @param [out] pNewSection : new section returned
+template <class Elf>
+void ElfWriter<Elf>::createNewSection(const char *pName, SectionBuffer *pInputSection, SectionBuffer *pNewSection) {
+
+  // Clear the new section
+  memset(pNewSection, 0, sizeof(SectionBuffer));
+
+  pNewSection->secHead = pInputSection->secHead;
+
+  // Add the section name to .strtab section
+  unsigned strTabSize = m_sections[m_strtabSecIdx].secHead.sh_size;
+  unsigned secNameSize = strlen(const_cast<char *>(pName)) + 1;
+
+  pNewSection->secHead.sh_name = strTabSize;
+  pNewSection->name = const_cast<char *>(pName);
+
+  pNewSection->secHead.sh_link = m_symSecIdx;
+  pNewSection->secHead.sh_size = 0;
+
+  auto strTabData = new uint8_t[strTabSize + secNameSize];
+  memcpy(strTabData, m_sections[m_strtabSecIdx].data, strTabSize);
+  memcpy(strTabData + strTabSize, pName, secNameSize);
+
+  delete[] m_sections[m_strtabSecIdx].data;
```

Comment:
Could we use `SmallVector<uint8_t> strTabData(strTabSize + secNameSize);`?

Reply:
It seems that this style is used everywhere in the `llpcElfWriter.cpp`. Do you  think we can make a separate commit to modify all of them?

Reply:
Sure